dialogue-core/src/main/java/com/palantir/dialogue/core/NodeSelectionStrategyChannel.java

```diff
@@ -0,0 +1,191 @@
+                    PinUntilErrorNodeSelectionStrategyChannel previousPinUntilError =
+                            (PinUntilErrorNodeSelectionStrategyChannel) previousNodeSelectionStrategy;
+                    return ChannelWithStrategy.of(
+                            updatedStrategy,
+                            PinUntilErrorNodeSelectionStrategyChannel.of(
+                                    Optional.of(previousPinUntilError.getCurrentChannel()),
+                                    updatedStrategy,
+                                    channels,
+                                    pinuntilerrorMetrics,
+                                    random,
+                                    channelName));
+                }
+                return ChannelWithStrategy.of(
+                        updatedStrategy,
+                        PinUntilErrorNodeSelectionStrategyChannel.of(
+                                Optional.empty(),
+                                updatedStrategy,
+                                channels,
+                                pinuntilerrorMetrics,
+                                random,
+                                channelName));
+            case BALANCED:
+                // When people ask for 'ROUND_ROBIN', they usually just want something to load balance better.
+                // We used to have a naive RoundRobinChannel, then tried RandomSelection and now use this heuristic:
+                return ChannelWithStrategy.of(
+                        updatedStrategy,
+                        new BalancedNodeSelectionStrategyChannel(channels, random, tick, metrics, channelName));
+            case UNKNOWN:
+        }
+        throw new SafeRuntimeException("Unknown NodeSelectionStrategy", SafeArg.of("unknown", updatedStrategy));
+    }
+
+    @Value.Immutable
+    interface ChannelWithStrategy {
+        DialogueNodeSelectionStrategy strategy();
+
+        LimitedChannel channel();
+
+        static ChannelWithStrategy of(DialogueNodeSelectionStrategy strategy, LimitedChannel channel) {
+            return ImmutableChannelWithStrategy.builder()
+                    .strategy(strategy)
+                    .channel(channel)
+                    .build();
+        }
+    }
+
+    private class WrapperChannel implements LimitedChannel {
+        private final LimitedChannel delegate;
+        private final FutureCallback<Response> callback;
+
+        WrapperChannel(LimitedChannel delegate) {
+            this.delegate = delegate;
+            this.callback = new FutureCallback<Response>() {
+                @Override
+                public void onSuccess(Response result) {
+                    result.getFirstHeader(NODE_SELECTION_HEADER)
+                            .ifPresent(strategy -> updateRequestedStrategies(
+                                    delegate, DialogueNodeSelectionStrategy.fromHeader(strategy)));
```

Comment:
Can we extract the reused lambdas? The magbeExecute result below can avoid the response-> addDirectCallback allocation as well.

The callback and transformer registrations showed up pretty heavily in the profiler when I took a quick look at our simulation performance, long term we might want to do something about that.